src/bitcoind/actions.rs

```diff
@@ -0,0 +1,138 @@
+    config_network: &String,
+) -> Result<(), BitcoindError> {
+    let chaininfo = bitcoind.getblockchaininfo()?;
+    let chain = chaininfo
+        .get("chain")
+        .and_then(|c| c.as_str())
+        .ok_or_else(|| {
+            BitcoindError("No valid 'chain' in getblockchaininfo response?".to_owned())
+        })?;
+
+    if !config_network.eq(chain) {
+        return Err(BitcoindError(format!(
+            "Wrong network, bitcoind is on '{}' but our config says '{}'",
+            chain, config_network
+        )));
+    }
+
+    Ok(())
+}
+
+/// Some sanity checks to be done at startup to make sure our bitcoind isn't going to fail under
+/// our feet for a legitimate reason.
+pub fn bitcoind_sanity_checks(
+    bitcoind: &BitcoinD,
+    bitcoind_config: &BitcoindConfig,
+) -> Result<(), BitcoindError> {
+    check_bitcoind_network(&bitcoind, &bitcoind_config.network)
+}
+
+/// Polls bitcoind until we are synced.
+/// Tries to be smart with getblockchaininfo calls.
+pub fn wait_for_bitcoind_synced(
+    bitcoind: &BitcoinD,
+    bitcoind_config: &BitcoindConfig,
+) -> Result<(), BitcoindError> {
+    // We need to take the edge case in which all headers aren't downloaded yet
+    // into account.
+    let mut first = true;
+
+    loop {
+        let chaininfo = bitcoind.getblockchaininfo()?;
+        let (headers, blocks, ibd) = (
+            chaininfo
+                .get("headers")
+                .and_then(|h| h.as_u64())
+                .ok_or_else(|| {
+                    BitcoindError("No valid 'headers' in getblockchaininfo response?".to_owned())
+                })?,
+            chaininfo
+                .get("blocks")
+                .and_then(|b| b.as_u64())
+                .ok_or_else(|| {
+                    BitcoindError("No valid 'blocks' in getblockchaininfo response?".to_owned())
+                })?,
+            chaininfo
+                .get("initialblockdownload")
+                .and_then(|i| i.as_bool())
+                .ok_or_else(|| {
+                    BitcoindError(
+                        "No valid 'initialblockdownload' in getblockchaininfo response?".to_owned(),
+                    )
+                })?,
+        );
+        let mut delta = if headers > blocks {
+            headers - blocks
+        } else {
+            0
+        };
+
+        if ibd {
+            // Ok, so we have some time. Let's try to avoid slowing it down by
+            // spamming it with getblockchaininfo calls.
+
+            if first {
+                log::info!(
+                    "Bitcoind is currently performing IBD, this is going to \
+                        take some time."
+                );
+            }
+
+            // First: wait for it to gather all headers, if the current delta is
+            // big enough. let's assume it won't take longer than 5min from now
+            // for mainnet.
+            // Then: get the number of blocks left to DL
+            if first || delta > 1_000 {
```

Comment:
move 1_000, 3 and 5*60 as constants

Reply:
Meh

That's kind of a hack that is going to be tinkered again, i don't even know how to name those constants.. `const DIFF_HEADERS_BLOCKS_IS_SMALL_ENOUGH = 1_000 // <- not sure about this`